app/knights/knight.py

```diff
@@ -0,0 +1,40 @@
+class Knight:
+    def __init__(self, knight: dict):
+        self.name = knight["name"]
+        self.power = knight["power"]
+        self.hp = knight["hp"]
+        self.armour = knight["armour"]
+        self.weapon = knight["weapon"]
+        self.potion = knight["potion"]
+        self.protection = 0
+
+    def battle_preparation(self):
+        # apply armour
+        self.protection = 0
+        for a in self.armour:
+            self.protection += a["protection"]
+
+        # apply weapon
+        self.power += self.weapon["power"]
+
+        # apply potion if exist
+        if self.potion is not None:
+            if "power" in self.potion["effect"]:
+                self.power += self.potion["effect"]["power"]
+
+            if "protection" in self.potion["effect"]:
+                self.protection += self.potion["effect"]["protection"]
+
+            if "hp" in self.potion["effect"]:
+                self.hp += self.potion["effect"]["hp"]
```

Comment:
These code duplications can be also fixed, using `stats = ("protection", "power", "hp")` and looping on `stats`